core/src/main/java/org/apache/iceberg/ManifestFilterManager.java

```diff
@@ -510,16 +510,19 @@ private Pair<InclusiveMetricsEvaluator, StrictMetricsEvaluator> metricsEvaluator
       // in other words, ResidualEvaluator returns a part of the expression that needs to be
       // evaluated
       // for rows in the given partition using metrics
-      return metricsEvaluators.computeIfAbsent(
-          file.partition(),
-          partition -> {
-            Expression residual = residualEvaluator.residualFor(partition);
-            InclusiveMetricsEvaluator inclusive =
-                new InclusiveMetricsEvaluator(tableSchema, residual, caseSensitive);
-            StrictMetricsEvaluator strict =
-                new StrictMetricsEvaluator(tableSchema, residual, caseSensitive);
-            return Pair.of(inclusive, strict);
-          });
+      PartitionData partition = (PartitionData) file.partition();
+      if (!metricsEvaluators.containsKey(partition)) {
```

Comment:
Will the fix still work if we just keep computeIfAbsent, but pass file.partition().copy()?

Reply:
Yes but then we make a copy of every single partition object, so we basically lose all benefit from reuse containers

Reply:
OK makes sense